forestflow/pcross.py

```diff
@@ -0,0 +1,233 @@
+        z (float): single redshift to evaluate
+    Optional Parameters:
+        P3D_mode: 'pol' or 'cart' for polar or cartesian. 'pol' assumes that the function takes parameters z and an array of k and mu. 'cart' assumes that the parameters are z, kpar and kperp, both arrays.
+        fast (bool): if true, the transition to P1D is done faster, without interpolation, but there will be a discontinuity
+        **P3D_kwargs: optional named arguments to be passed to the P3D function.
+    Returns:
+        rperp, Px_per_rperp, Px_per_kpar
+        rperp: array of log-space r-perpendicular (separation in Mpc)
+        Px_per_kpar: P-cross as an array with shape (len(kpars), len(rperp)).
+    """
+    import hankl
+    from scipy.interpolate import CubicSpline
+    nkperp = 2**16
+    nkpar  = len(kpars)
+    kperps = np.logspace(
+        np.log10(10.0**-20), np.log10(10.0**3), nkperp
+    )  # set up an array of kperp
+    # tile
+    kperp2d  = np.tile(kperps[:, np.newaxis], nkpar) # mu grid for P3D
+    kpar2d   = np.tile(kpars[:, np.newaxis], nkperp).T
+    k2d = np.sqrt(kperp2d**2 + kpar2d**2)
+    mu2d = kpar2d / k2d
+    if P3D_mode == 'cart':
+        # assume P3D is a function of (z, kpar, kperp)
+        P3D_eval = P3D(z=z, kpar=kpar2d, kperp=kperp2d, **P3D_kwargs)
+    elif P3D_mode == 'pol':
+        # assume P3D is a function of (z, k, mu)
+        P3D_eval = P3D(z=z, k=k2d, mu=mu2d, **P3D_kwargs)
+    P1D = P1D_Mpc(P3D, z, np.linspace(np.log(0.001), np.log(100), 99), kpars, P3D_mode = P3D_mode, **P3D_kwargs) # get P1D
+    Px_per_kpar = []
+    print(P3D_eval.shape)
+    for ik, kpar in enumerate(kpars):  # for each value of k parallel to evaluate Px at
+        P3D_kpar = P3D_eval[:,ik] # get the P3D
+        func     = P3D_kpar * kperps
+        rperp, LHS = hankl.FFTLog(
+            kperps, func, q=0, mu=0
+        )  # returns an array of log-spaced rperps, and the Hankel Transform
+        Px = LHS / rperp / (2*np.pi)  # Divide out by remaining factor to get Px
+        # transition    
+        # replace the values left of the minimum
+        replace = rperp < 0.02
+        # return the P1D result for that kpar
+        Px[replace] = P1D[ik]
+        # between rperp = 0.02 and 0.08, interpolate from P1D to Px values
+        idxmin = (np.abs(rperp - 0.02)).argmin()
+        idxmax = (np.abs(rperp - 0.08)).argmin()
+        rperps_interp = rperp[idxmin:idxmax]
+        Px_tointerp = np.delete(Px, np.arange(idxmin, idxmax))
+        rperp_tointerp = np.delete(rperp, np.arange(idxmin, idxmax))
+        interpmin = np.abs(rperp_tointerp - 0.005).argmin()
+        interpmax = np.abs(rperp_tointerp - 0.2).argmin()
+        cs = CubicSpline(
+            rperp_tointerp[interpmin:interpmax],
+            Px_tointerp[interpmin:interpmax],
+        )
+        Px_interpd = cs(rperps_interp)
+        Px = np.insert(Px_tointerp, idxmin, Px_interpd)
+        min_rperp, max_rperp = 10**-2, 100
+        if min_rperp > min(rperp):
```

Comment:
It would be great to put a warning.warn() explaining that you are changing the output range required by the sure, right? (I may have got confused though)

Reply:
These min and max rperp are user options in the detailed Px function, but not in this version of the function for simplicity. The next commit will have some more documentation on the fact that the function auto-returns a range from .01 to 100 if specific rperp values are not requested.